test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc

```diff
@@ -0,0 +1,279 @@
+#include "test/extensions/transport_sockets/tls/rsa_private_key_method_provider.h"
+
+#include <memory>
+
+#include "envoy/api/api.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace PrivateKeyMethodProvider {
+
+int RsaPrivateKeyMethodProvider::ssl_rsa_connection_index = -1;
+
+void RsaPrivateKeyConnection::delayed_op() {
+  const std::chrono::milliseconds timeout_0ms{0};
+
+  timer_ = dispatcher_.createTimer([this]() -> void {
+    finished_ = true;
+    this->cb_.complete();
+    return;
+  });
+  timer_->enableTimer(timeout_0ms);
+}
+
+static ssl_private_key_result_t privateKeySign(SSL* ssl, uint8_t* out, size_t* out_len,
+                                               size_t max_out, uint16_t signature_algorithm,
+                                               const uint8_t* in, size_t in_len) {
+  RSA* rsa;
+  bssl::ScopedEVP_MD_CTX ctx;
+  const EVP_MD* md;
+  RsaPrivateKeyConnection* ops = static_cast<RsaPrivateKeyConnection*>(
+      SSL_get_ex_data(ssl, RsaPrivateKeyMethodProvider::ssl_rsa_connection_index));
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int hash_len;
+  uint8_t* msg;
+  size_t msg_len;
+  int prefix_allocated = 0;
+  int padding = RSA_NO_PADDING;
+
+  if (!ops) {
+    return ssl_private_key_failure;
+  }
+
+  if (ops->test_options_.method_error_) {
+    // Have an artificial test failure.
+    return ssl_private_key_failure;
+  }
+
+  if (!ops->test_options_.sign_expected_) {
+    // TODO(ipuustin): throw exception, because a failure can be an expected result in some tests?
+    return ssl_private_key_failure;
+  }
+
+  rsa = ops->getPrivateKey();
+  if (rsa == nullptr) {
+    return ssl_private_key_failure;
+  }
+
+  md = SSL_get_signature_algorithm_digest(signature_algorithm);
+  if (!md) {
+    return ssl_private_key_failure;
+  }
+
+  // Calculate the digest for signing.
+  if (!EVP_DigestInit_ex(ctx.get(), md, nullptr) || !EVP_DigestUpdate(ctx.get(), in, in_len) ||
+      !EVP_DigestFinal_ex(ctx.get(), hash, &hash_len)) {
+    return ssl_private_key_failure;
+  }
+
+  // Add RSA padding to the the hash.
+  if (SSL_is_signature_algorithm_rsa_pss(signature_algorithm)) {
+    msg_len = RSA_size(rsa);
+    msg = static_cast<uint8_t*>(OPENSSL_malloc(msg_len));
```

Comment:
(The comments from the ECDSA code largely apply here too.)

Reply:
Check.